### PR TITLE
Studio: Allow blocking user from adding demo sites 

### PR DESCRIPTION
**Pull request by katinthehatsite**
Related to https://github.com/Automattic/dotcom-forge/issues/7370

## Proposed Changes

This PR implements the UI for when the user is blocked from creating a demo site:

Screenshot 2024-07-11 at 9 10 57 PM

## Testing Instructions

* Navigate to the User Report Card for your own user
* Use the `Disable site creation` button (at the bottom of the User Report Card) to block your user from creating demo sites
* Pull the changes from this branch
* Start the app with `nvm use && npm install && npm start`
* Make sure to log in with the user that you just blocked from creating sites
* Navigate to `Share` tab
* Confirm that `Add demo site` button is disabled
* Hover over the button and confirm that it displays a message about demo sites being disabled for this account

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/hooks/use-archive-error-messages.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export function useArchiveErrorMessages() {
 				rest_cannot_view: __(
 					"There's been an authentication error. Please log in again before sharing a site."
 				),
+				rest_site_creation_blocked: __( 'Demo sites are not available for your account.' ),
```
Comment (katinthehatsite) [inline src/hooks/use-archive-error-messages.ts:23]: Not directly related to this PR but I thought it would be good to add this error in the list of the specific errors that we have

**Comment by wojtekn** [comment]
@katinthehatsite could we also account for a case in which we disable demo sites for the user who has Studio already open? Currently, when such a user tries to add a demo site, they will get a proper error message, but the button won't be disabled:

Screenshot 2024-07-12 at 12 22 03

**Comment by katinthehatsite** [comment]
Hmm let me check what is happening there 👀

**Comment by katinthehatsite** [comment]
@wojtekn I made some updates, would you be able to give it another try?

File: src/hooks/use-snapshots.tsx [hunk 2]
```diff
@@ -280,13 +284,16 @@ export const SnapshotProvider: React.FC< { children: ReactNode } > = ( { childre
...
+			const response: { site_count: number; site_limit: number; site_creation_blocked: boolean } =
+				await client.req.get( {
+					path: '/jurassic-ninja/usage',
+					apiNamespace: 'wpcom/v2',
+				} );
+			setSnapshotCreationBlocked( response.site_creation_blocked );
 			return response;
 		} catch ( error ) {
 			Sentry.captureException( error );
+			setSnapshotCreationBlocked( false ); // Default the site creation to enabled
```
Comment (katinthehatsite) [inline src/hooks/use-snapshots.tsx]: Not sure if in case of an error when getting the data from the endpoint, we should keep the creation disabled or enabled. My thinking was to set it to enabled regardless so that the user could attempt the action and in the worst case scenario, they would get a dialog window with an error.

Happy to change it though if there is a better approach

File: src/components/content-tab-snapshots.tsx [hunk 3]
```diff
@@ -390,14 +396,28 @@ function AddDemoSiteWithProgress( {
 		interval: 1500,
 		maxValue: 95,
 	} );
+
+	useEffect( () => {
+		const intervalId = setInterval( () => {
+			fetchSnapshotUsage(); // Function to fetch the latest data from the endpoint
+		}, 10000 );
```
Comment (katinthehatsite) [inline src/components/content-tab-snapshots.tsx]: The only way I found to get the real-time data was to poll the endpoint - not sure if it is the best solution. Perhaps we can reduce the polling interval as the case of the user being blocked does not seem to be something that would happen often. This would help to not send as many requests 😅

Comment (fluiddot) [reply src/components/content-tab-snapshots.tsx]: Yeah, maybe we could consider other options to avoid making requests consistently. I'd propose checking usage (i.e. making the request to the AP) in the following cases:
1. Opening the app.
2. Creating/updating a demo site, before archiving the site to bail quickly ([reference](https://github.com/Automattic/studio/blob/5b834ea553f5097dc9caeee0fb6baecb4994ba52/src/hooks/use-archive-site.ts#L65-L70)).

We could also do the check when the app gains focus, but as you pointed out @katinthehatsite, blocking users won't be common so we can discard this case.

File: src/hooks/use-snapshots.tsx [hunk 4]
```diff
@@ -280,13 +284,16 @@ export const SnapshotProvider: React.FC< { children: ReactNode } > = ( { childre
 		}
 		setIsLoading( true );
 		try {
-			const response: { site_count: number; site_limit: number } = await client.req.get( {
-				path: '/jurassic-ninja/usage',
-				apiNamespace: 'wpcom/v2',
-			} );
+			const response: { site_count: number; site_limit: number; site_creation_blocked: boolean } =
```
Comment (fluiddot) [inline src/hooks/use-snapshots.tsx:289]: We could also add the property `site_creation_blocked` to the type definition of the function `fetchSnapshotUsage`.

https://github.com/Automattic/studio/blob/5b834ea553f5097dc9caeee0fb6baecb4994ba52/src/hooks/use-snapshots.tsx#L36

File: src/components/content-tab-snapshots.tsx (hunk 3 above)
Comment (fluiddot) [reply src/components/content-tab-snapshots.tsx]: Here's a potential approach:

```patch
diff --git forkSrcPrefix/src/hooks/use-snapshots.tsx forkDstPrefix/src/hooks/use-snapshots.tsx
index 564090ea621e4144ab68922c4083ffb66cb17cb8..32abc542d3d40f057290ac7c044c6c1ffc270587 100644
--- forkSrcPrefix/src/hooks/use-snapshots.tsx
+++ forkDstPrefix/src/hooks/use-snapshots.tsx
@@ -33,7 +33,9 @@ interface SnapshotContextType {
 	loadingServerSnapshots: boolean;
 	activeSnapshotCount: number;
 	snapshotQuota: number;
-	fetchSnapshotUsage: () => Promise< { site_limit: number; site_count: number } | undefined >;
+	fetchSnapshotUsage: () => Promise<
+		{ site_limit: number; site_count: number; site_creation_blocked: boolean } | undefined
+	>;
 	isLoadingSnapshotUsage: boolean;
 	initiated: boolean;
 	snapshotCreationBlocked: boolean;
diff --git forkSrcPrefix/src/components/content-tab-snapshots.tsx forkDstPrefix/src/components/content-tab-snapshots.tsx
index 8a1d8c3571b78a26e85868f0b1dc57fe1fa61be8..3ffa66b760599f11da7d755f82bb40c925d0b93a 100644
--- forkSrcPrefix/src/components/content-tab-snapshots.tsx
+++ forkDstPrefix/src/components/content-tab-snapshots.tsx
@@ -381,13 +381,8 @@ function AddDemoSiteWithProgress( {
 	const { __, _n } = useI18n();
 	const { archiveSite, isUploadingSiteId, isAnySiteArchiving } = useArchiveSite();
 	const isUploading = isUploadingSiteId( selectedSite.id );
-	const {
-		activeSnapshotCount,
-		snapshotQuota,
-		isLoadingSnapshotUsage,
-		snapshotCreationBlocked,
-		fetchSnapshotUsage,
-	} = useSnapshots();
+	const { activeSnapshotCount, snapshotQuota, isLoadingSnapshotUsage, snapshotCreationBlocked } =
+		useSnapshots();
 	const isLimitUsed = activeSnapshotCount >= snapshotQuota;
 	const isOffline = useOffline();
 	const { progress, setProgress } = useProgressTimer( {
@@ -397,14 +392,6 @@ function AddDemoSiteWithProgress( {
 		maxValue: 95,
 	} );
 
-	useEffect( () => {
-		const intervalId = setInterval( () => {
-			fetchSnapshotUsage(); // Function to fetch the latest data from the endpoint
-		}, 10000 );
-
-		return () => clearInterval( intervalId ); // Cleanup the interval on component unmount
-	}, [ fetchSnapshotUsage ] );
-
 	useEffect( () => {
 		if ( isSnapshotLoading ) {
 			setProgress( 80 );
diff --git forkSrcPrefix/src/hooks/use-archive-site.ts forkDstPrefix/src/hooks/use-archive-site.ts
index 02061b044032c14fc4fcfc001828eaae465107c1..6d0d5795cc0073d6121f0feb59e8d2db8e4f97b5 100644
--- forkSrcPrefix/src/hooks/use-archive-site.ts
+++ forkDstPrefix/src/hooks/use-archive-site.ts
@@ -12,7 +12,7 @@ import { SnapshotStatus, SnapshotStatusResponse, useSnapshots } from './use-snap
 
 export function useArchiveSite() {
 	const { uploadingSites, setUploadingSites } = useSiteDetails();
-	const { snapshots, addSnapshot, updateSnapshot } = useSnapshots();
+	const { snapshots, addSnapshot, updateSnapshot, fetchSnapshotUsage } = useSnapshots();
 	const isUploadingSiteId = useCallback(
 		( localSiteId: string ) => uploadingSites[ localSiteId ] || false,
 		[ uploadingSites ]
@@ -70,6 +70,14 @@ export function useArchiveSite() {
 			}
 
 			setUploadingSites( ( _uploadingSites ) => ( { ..._uploadingSites, [ siteId ]: true } ) );
+
+			const snapshotUsage = await fetchSnapshotUsage();
+			if ( snapshotUsage?.site_creation_blocked ) {
+				alert( errorMessages.rest_site_creation_blocked );
+				setUploadingSites( ( _uploadingSites ) => ( { ..._uploadingSites, [ siteId ]: false } ) );
+				return;
+			}
+
 			const { zipContent, zipPath, exceedsSizeLimit } = await getIpcApi().archiveSite( siteId );
 			if ( exceedsSizeLimit ) {
 				alert(

```

File: src/components/content-tab-snapshots.tsx [hunk 5]
```diff
@@ -410,6 +430,7 @@ function AddDemoSiteWithProgress( {
 		snapshotQuota
 	);
 	const offlineMessage = __( 'Creating a demo site requires an internet connection.' );
+	const userBlockedMessage = __( 'Demo sites are not available for your account.' );
```
Comment (fluiddot) [inline src/components/content-tab-snapshots.tsx]: I wonder if we could reuse the string from `useArchiveErrorMessages` hook:
https://github.com/Automattic/studio/blob/5b834ea553f5097dc9caeee0fb6baecb4994ba52/src/hooks/use-archive-error-messages.ts#L23

**Review by fluiddot — COMMENTED** [review]
I noticed that if the user being blocked already has demo sites created, the Update demo site button is enabled. I'd apply the same approach we've done for the Add button there.

Screenshot 2024-07-18 at 10 56 51



Additionally, when updating a demo site, I think the error message for blocked users should mention the failure reason

Screenshot 2024-07-18 at 10 37 17

File: src/hooks/use-snapshots.tsx (hunk 2 above)
Comment (fluiddot) [reply src/hooks/use-snapshots.tsx]: Probably we don't need to change the value when the request fails, as we don't know the value to use. Ultimately, if the user is blocked, the site won't be created on the server's end.

File: src/components/content-tab-snapshots.tsx (hunk 3 above)
Comment (fluiddot) [reply src/components/content-tab-snapshots.tsx]: The main downside of the above approach compared to polling the API would be for the case of blocked users being unblocked. As the demo site creation won't be enabled until they restart the app. In any case, I presume this will be an uncommon case that we could omit in the implementation. Alternatively, we could check usage when the app gains focus.

File: src/hooks/use-snapshots.tsx (hunk 2 above)
Comment (katinthehatsite) [reply src/hooks/use-snapshots.tsx]: Sounds good, I made chanegs in https://github.com/Automattic/studio/pull/354/commits/bba74b57a5120e21952eba668971e210eb2e3ee5

File: src/hooks/use-snapshots.tsx (hunk 4 above)
Comment (katinthehatsite) [reply src/hooks/use-snapshots.tsx:289]: Added in https://github.com/Automattic/studio/pull/354/commits/51eda6825b2d8ca60bb6aa07849dc8ae7dce1748

File: src/components/content-tab-snapshots.tsx (hunk 5 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx]: I fixed that in https://github.com/Automattic/studio/pull/354/commits/0b95433de9d3b9d7107a3225448ac7d57a4aea4f

**Comment by katinthehatsite** [comment]
Thanks for the review @fluiddot ! I made some changes to block the `Update site` button when the user is blocked in https://github.com/Automattic/studio/pull/354/commits/0b95433de9d3b9d7107a3225448ac7d57a4aea4f

I tried the alternative solution but I think at the end, it comes to the same thing that @wojtekn mentioned initially with the app not updating the data in real-time. I left it as is for now, let's wait for some comments and see if we should change the approach of polling the endpoint.

File: src/components/content-tab-snapshots.tsx [hunk 6]
```diff
@@ -182,16 +193,22 @@ function SnapshotRow( {
 				) : (
 					<>
 						<Tooltip
-							disabled={ ! isOffline }
-							icon={ offlineIcon }
-							text={ updateDemoSiteOfflineMessage }
+							disabled={ ! ( isOffline || snapshotCreationBlocked ) }
+							{ ...( isOffline && { icon: offlineIcon } ) }
+							text={ isOffline ? updateDemoSiteOfflineMessage : blockedUserMessage }
```
Comment (sejas) [inline src/components/content-tab-snapshots.tsx]: What do you think about using a new constant for the props? Similar to the Add demo site tootltip?
I've created this PR with my suggestion: https://github.com/Automattic/studio/pull/388

File: src/components/content-tab-snapshots.tsx [hunk 7]
```diff
@@ -182,16 +193,22 @@ function SnapshotRow( {
...
+										? blockedUserMessage
+										: isOffline
+										? updateDemoSiteOfflineMessage
+										: ''
+								}
+								aria-disabled={ isOffline || snapshotCreationBlocked }
 								variant="primary"
 								onClick={ () => {
-									if ( isOffline ) {
+									if ( isOffline || snapshotCreationBlocked ) {
```
Comment (sejas) [inline src/components/content-tab-snapshots.tsx]: In my suggested PR I'm moving this condition into a new variable, because it's repeated multiple times.

File: src/components/content-tab-snapshots.tsx [hunk 8]
```diff
@@ -421,6 +447,8 @@ function AddDemoSiteWithProgress( {
 		tooltipContent = { text: allotmentConsumptionMessage };
 	} else if ( isAnySiteArchiving ) {
 		tooltipContent = { text: siteArchivingMessage };
+	} else if ( snapshotCreationBlocked ) {
+		tooltipContent = { text: userBlockedMessage };
```
Comment (sejas) [inline src/components/content-tab-snapshots.tsx:457]: What about assigning the message directly?

```suggestion
		tooltipContent = { text: errorMessages.rest_site_creation_blocked };
```

File: src/hooks/use-snapshots.tsx [hunk 9]
```diff
@@ -280,10 +286,12 @@ export const SnapshotProvider: React.FC< { children: ReactNode } > = ( { childre
...
+				await client.req.get( {
+					path: '/jurassic-ninja/usage',
+					apiNamespace: 'wpcom/v2',
+				} );
+			setSnapshotCreationBlocked( response.site_creation_blocked );
```
Comment (sejas) [inline src/hooks/use-snapshots.tsx:294]: What do you think about moving `setActiveSnapshotCount` and  `setSnapshotQuota` from `fetchStats` to here ?

File: src/components/content-tab-snapshots.tsx [hunk 10]
```diff
@@ -70,6 +73,14 @@ function SnapshotRow( {
 		maxValue: 95,
 	} );
 
+	useEffect( () => {
+		const intervalId = setInterval( () => {
+			fetchSnapshotUsage(); // Function to fetch the latest data from the endpoint
+		}, 10000 );
+
+		return () => clearInterval( intervalId ); // Cleanup the interval on component unmount
+	}, [ fetchSnapshotUsage ] );
```
Comment (sejas) [inline src/components/content-tab-snapshots.tsx]: I would avoid using an active listening. It not "wrong" and it works as expected, but I see it unnecessary and a trend that could  lead in adding more listeners all the time.

Maybe one request when the component mounts is enough?

```suggestion
	useEffect( () => {
		fetchSnapshotUsage();
	}, [ fetchSnapshotUsage ] );
```

My reasoning is that the backend will fail anyway and after it fails we can update the status of this button and disable it with the correct tooltip.

**Review by sejas — COMMENTED** [review]
I've created a PR to simplify the logic of the update demo site button. Let me know what do you think. https://github.com/Automattic/studio/pull/388

I've tested login in, loging out, enabling and disabling my user RC setting, and it works as expected. Great job!

https://github.com/user-attachments/assets/3a802eb3-9844-4f02-8888-04ae8e7fca1a

From my point of view I would avoid adding the active listening. I see it as a corner case.
I like how we refresh the state when the user  starts the app or logs in. We should use a similar criteria as we do with the quota of demo sites.

Instead of adding an active listening, I would disable the button after the first fail. Not a strong opinion.

File: src/components/content-tab-snapshots.tsx (hunk 6 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx]: Yeah, I think it is a good idea. It makes the code more readable. I merged the PR into this PR 👍

File: src/components/content-tab-snapshots.tsx (hunk 8 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx:457]: Yeah that's what I had initially https://github.com/Automattic/studio/pull/354/commits/5e6720ef8fe45640aecdf70b9867501611c4d6d2 but in one feedback comments, it was mentioned that we could pull it from the `errorMessages`. Let's go back to using the string directly to be consistent with other tooltip messages.

Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx:457]: Updated in https://github.com/Automattic/studio/pull/354/commits/e0ac36b0a0afd1c4be4ce79a780e760b1ff0f6b1

File: src/components/content-tab-snapshots.tsx (hunk 10 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx]: Yeah that was my initial thinking; I have made these changes in https://github.com/Automattic/studio/pull/354/commits/fbd83095550af6a441a261c742cb799f9841bad5

File: src/hooks/use-snapshots.tsx (hunk 9 above)
Comment (katinthehatsite) [reply src/hooks/use-snapshots.tsx:294]: Actually initially I have done the reverse and added `setSnapshotCreationBlocked` to `fetchStats` but when refactoring for polling the endpoint, I think this made more sense. 

Generally, I don't have a strong preference but one thing I am worried about is that these changes might have been intentional (meaning for fetching stats), and I would be worried about breaking something that does not necessarily directly relate to the fix in this PR. I would prefer to leave this as is and not touch it.

File: src/components/content-tab-snapshots.tsx (hunk 7 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx]: Makes sense, thanks. I have merged that PR so I will keep your approach 👍

**Comment by katinthehatsite** [comment]
> From my point of view I would avoid adding the active listening. I see it as a corner case.
I like how we refresh the state when the user starts the app or logs in. We should use a similar criteria as we do with the quota of demo sites.

Based on our discussions of this issue, I have implemented the behaviour mentioned in the comment above. Since it is more of an edge case, I would be okay not to poll the endpoint and update the information in real-time.

@sejas could you please give it another test and let me know what you think? Thanks!

File: src/components/content-tab-snapshots.tsx (hunk 8 above)
Comment (sejas) [reply src/components/content-tab-snapshots.tsx:457]: @katinthehatsite , Sorry I meant to use the `errorMessages.rest_site_creation_blocked` directly without creating an intermediary variable.

File: src/hooks/use-snapshots.tsx (hunk 9 above)
Comment (sejas) [reply src/hooks/use-snapshots.tsx:294]: Makes sense, initially I was also thinking about moving `setSnapshotCreationBlocked` to `fetchStats`, but fetchStats is not exported and it would require a bit of refactor. It's ok to leave it as it is.

File: src/components/content-tab-snapshots.tsx (hunk 8 above)
Comment (katinthehatsite) [reply src/components/content-tab-snapshots.tsx:457]: Alright, let's fix it up 👍

**Review by sejas — APPROVED** [review]
Kat!, thanks for applying the suggested cahnges.

I've tested the edge case when a user gets blocked and it has Studio running.
As shown in the video, the create demo site will fail and immediately will see the tooltip 👍 .

https://github.com/user-attachments/assets/fadedf32-d522-4931-8d87-43078e30b3b4

**Comment by fluiddot** [comment]
> Thanks for the review @fluiddot ! I made some changes to block the `Update site` button when the user is blocked in [0b95433](https://github.com/Automattic/studio/commit/0b95433de9d3b9d7107a3225448ac7d57a4aea4f)

Thanks @katinthehatsite for addressing the comments I shared. I've tested the changes in `trunk` and confirmed it worked as expected 🎊 .
